core/src/syscall/precompiles/sha512/compress/air.rs

```diff
@@ -0,0 +1,484 @@
+use core::borrow::Borrow;
+
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_field::AbstractField;
+use p3_matrix::Matrix;
+
+use super::{
+    columns::{Sha512CompressCols, NUM_SHA512_COMPRESS_COLS},
+    Sha512CompressChip,
+};
+use crate::{
+    air::{
+        AluAirBuilder, BaseAirBuilder, ByteAirBuilder, MemoryAirBuilder, Word64, WordAirBuilder,
+    },
+    bytes::ByteOpcode,
+    memory::MemoryCols,
+    operations::{
+        Add64Operation, And64Operation, FixedRotateRight64Operation, Not64Operation, Xor64Operation,
+    },
+    runtime::SyscallCode,
+};
+
+impl<F> BaseAir<F> for Sha512CompressChip {
+    fn width(&self) -> usize {
+        NUM_SHA512_COMPRESS_COLS
+    }
+}
+
+impl<AB> Air<AB> for Sha512CompressChip
+where
+    AB: BaseAirBuilder,
+{
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let (local, next) = (main.row_slice(0), main.row_slice(1));
+        let local: &Sha512CompressCols<AB::Var> = (*local).borrow();
+        let next: &Sha512CompressCols<AB::Var> = (*next).borrow();
+
+        // Constrain the incrementing nonce.
+        builder.when_first_row().assert_zero(local.nonce);
+        builder
+            .when_transition()
+            .assert_eq(local.nonce + AB::Expr::one(), next.nonce);
+
+        // Assert that is_real is a bool.
+        builder.assert_bool(local.is_real);
+
+        self.eval_memory(builder, local);
+
+        self.eval_compression_ops(builder, local);
+
+        builder.receive_syscall(
+            local.shard,
+            local.channel,
+            local.clk,
+            local.nonce,
+            AB::F::from_canonical_u32(SyscallCode::SHA512_COMPRESS.syscall_id()),
+            local.w_ptr,
+            local.h_ptr,
+            local.is_real,
+        );
+    }
+}
+
+impl Sha512CompressChip {
+    /// Constrains that memory accesses are correct.
+    fn eval_memory<AB: MemoryAirBuilder>(
+        &self,
+        builder: &mut AB,
+        local: &Sha512CompressCols<AB::Var>,
+    ) {
+        // Assert `i` was read and written to correctly.
+        builder.eval_memory_access(
+            local.shard,
+            local.channel,
+            local.clk,
+            local.h_ptr + AB::F::from_canonical_u32(8 * 8),
+            &local.i_mem,
+            local.is_real,
+        );
+        let reduced_prev_i = local.i_mem.prev_value().reduce::<AB>();
+        builder
```

Comment:
We should be able to trust `i` since it's coming from the Rust code that calls `compress`